aten/src/ATen/native/ScatterGather.cpp

```diff
@@ -0,0 +1,85 @@
+#include <tuple>
+#include <algorithm>
+#include <ATen/ATen.h>
+#include <c10/core/WrapDimMinimal.h>
+
+namespace at { namespace native {
+
+Tensor & gather_out(Tensor & result, const Tensor & self, int64_t dim, const Tensor & index, bool sparse_grad) {
+  return at::_gather_out(result, self, dim, index, sparse_grad);
+}
+
+Tensor gather(const Tensor & self, int64_t dim, const Tensor & index, bool sparse_grad) {
+  return at::_gather(self, dim, index, sparse_grad);
+}
+
+Tensor & scatter_(Tensor & self, int64_t dim, const Tensor & index, const Tensor & source) {
+  TORCH_CHECK(self.is_contiguous(), "scatter_ only apply to contiguous tensor");
+  return at::_scatter_(self, dim, index, source);
+}
+
+Tensor & scatter_(Tensor & self, int64_t dim, const Tensor & index, Scalar value) {
+  TORCH_CHECK(self.is_contiguous(), "scatter_ only apply to contiguous tensor");
+  return at::_scatter_(self, dim, index, value);
+}
+
+Tensor scatter(const Tensor & self, int64_t dim, const Tensor & index, const Tensor & source) {
+  Tensor ret = self.clone().contiguous();
+  return at::_scatter_(ret, dim, index, source);
+}
+
+Tensor scatter(const Tensor & self, int64_t dim, const Tensor & index, Scalar value) {
+  Tensor ret = self.clone().contiguous();
+  return at::_scatter_(ret, dim, index, value);
+}
+
+Tensor & scatter_add_(Tensor & self, int64_t dim, const Tensor & index, const Tensor & source) {
+  TORCH_CHECK(self.is_contiguous(), "scatter_add_ only apply to contiguous tensor");
```

Comment:
we shouldn't throw an error in the forward because a backwards is broken, we should fix the backward to work with non-contiguous tensors.

Reply:
I think it is not just the backward is broken. For example, when the user writes 
```python
torch.tensor(0.0).expand(5, 5).scatter_(1, torch.eye(5, 5, dtype=torch.long), torch.ones(5, 5))
```
do they really mean to get a:
```python
tensor([[1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.]])
```
?
I think they are more interested in getting a
```python
tensor([[1., 1., 0., 0., 0.],
        [1., 1., 0., 0., 0.],
        [1., 1., 0., 0., 0.],
        [1., 1., 0., 0., 0.],
        [1., 1., 0., 0., 0.]])
```

Reply:
ah I see, you are concerned with partial aliasing.  See https://github.com/pytorch/pytorch/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+label%3A%22topic%3A+partial+aliasing%22+ for more information.

if you want to add error checking for partial aliasing, by all means go ahead!  @zou3519 or @pbelevich can help you.  But you are throwing out all non-contiguous tensors, many (most?) of which don't have memory overlap.